File: w5-array/Square.py
def dfs(length, finish, add, start, used, sticks, n):

    if finish == 3:  # 已完成3條邊，第4條邊自動完成
        return True

    if add == length:  # 當前邊已完成，開始下一條邊
        return dfs(length, finish + 1, 0, 0, used, sticks, n)

    for i in range(start, n):
        if not used[i] and add + sticks[i] <= length:
            used[i] = True
            if dfs(length, finish, add + sticks[i], i + 1, used, sticks, n):
                return True
            used[i] = False
        elif add + sticks[i] > length:
            break  # 剪枝：當前木棒太長，後面的更長

    return False


def can_form_square(sticks):
    total_sum = sum(sticks)
    max_stick = max(sticks)

    # 總和必須是4的倍數，且最長木棒不能超過邊長
    if total_sum % 4 != 0 or max_stick > total_sum // 4:
        return False

    target_length = total_sum // 4
    n = len(sticks)
    used = [False] * n

    sticks.sort(reverse=True)

    return dfs(target_length, 0, 0, 0, used, sticks, n)


def main():
    t = int(input())

    for _ in range(t):
        n = int(input())
        sticks = list(map(int, input().split()))

        if can_form_square(sticks):
            print("yes")
        else:
            print("no")


if __name__ == "__main__":
    main()

File: w5-array/Square.cpp
#include <iostream>
#include <algorithm>
using namespace std;

bool ok = false;
int MAP[50000] = {}, stick[50000] = {}, N;

void DFS(const int length, const int finish, const int add, const int start)
{
    if (finish == 3 || ok)
    {
        ok = true;
        return;
    }
    if (add == length)
    {
        DFS(length, finish + 1, 0, 0);
        return;
    }
    for (int i = start; i < N; i++)
    {
        if (MAP[i] == 0 && add + stick[i] <= length)
        {
            MAP[i] = 1;
            DFS(length, finish, add + stick[i], i + 1);
            MAP[i] = 0;
        }
        else if (stick[i] + add > length)
            return;
    }
}

int main()
{
    cin.sync_with_stdio(0);
    cin.tie(0);
    int T;
    cin >> T;
    for (int i = 0; i < T; i++)
    {
        int sum = 0, max = -999;
        cin >> N;
        ok = false;
        for (int j = 0; j < N; j++)
        {
            MAP[j] = 0;
        }
        for (int j = 0; j < N; j++)
        {
            cin >> stick[j];
            sum += stick[j];
            if (stick[j] > max)
                max = stick[j];
        }
        if (sum % 4 != 0 || max > sum / 4)
        {
            cout << "no\n";
            continue;
        }
        sort(stick, stick + N);
        DFS(sum / 4, 0, 0, 0);
        if (ok)
            cout << "yes\n";
        else
            cout << "no\n";
    }
}

// ZeroJudge D375
// Dr. SeanXD